### PR TITLE
[REVIEW] Port avro.pyx to libcudf++

### LINKED ISSUE #3920: [FEA] Port avro.pyx to use non-legacy libcudf APIs

### PR DESCRIPTION
**Pull request by trxcllnt**
Port `read_avro` Cython to new libcudf++ API.

Closes #3920.

### CODE REVIEW

File: python/cudf/cudf/_libxx/cpp/io/functions.pxd [hunk 1]
```diff
@@ -1,19 +1,25 @@
...
 from libcpp.string cimport string
 from libcpp.vector cimport vector
 
-cimport cudf._libxx.cpp.io.io_types as cudf_io_types
-
+from cudf._libxx.cpp.types cimport size_type
+from cudf._libxx.cpp.io.types cimport compression_type
+from cudf._libxx.cpp.io.types cimport quote_style
+from cudf._libxx.cpp.io.types cimport source_info
+from cudf._libxx.cpp.io.types cimport table_with_metadata
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/cpp/io/functions.pxd]:
```suggestion
from cudf._libxx.cpp.io.types cimport (
    compression_type,
    quote_style,
    source_info,
    table_with_metadata
)
```

File: python/cudf/setup.py [hunk 2]
```diff
@@ -80,7 +80,6 @@
     ext_modules=cythonize(
         extensions,
         nthreads=nthreads,
-        language="c++",
```
Comment (kkraus14) [inline python/cudf/setup.py]: We need to specify the language is `c++` somewhere I think. Maybe as a compiler directive?

Comment (trxcllnt) [reply python/cudf/setup.py]: I think this is covered by the `language="c++"` argument to the Extension constructor?

Comment (kkraus14) [reply python/cudf/setup.py]: I'm not sure. @jakirkham do you happen to know this?

Comment (kkraus14) [reply python/cudf/setup.py]: Nevermind this looks to be good: https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#specify-c-language-in-setup-py

Comment (jakirkham) [reply python/cudf/setup.py]: Yeah I think Paul's previous PR ensured we always use C++ in all Cython extensions.

File: cpp/include/cudf/io/functions.hpp [hunk 3]
```diff
@@ -52,6 +52,12 @@ struct read_avro_args {
   /// Rows to read; -1 is all
   size_type num_rows = -1;
 
+  read_avro_args() = default;
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply cpp/include/cudf/io/functions.hpp]: Yes, unfortunately cython needs a default constructor to allocate C++ structs on the stack. There's hacky workarounds, but they require heap allocation and manual deallocation, so it's easier to have a default constructor here.

**Pull request merged**